packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -63,40 +96,54 @@ export class BlobManager {
     private readonly blobIds: Set<string> = new Set();
     // blobs for which upload is pending. maps to a promise that will resolve once the blob has been uploaded and a
     // BlobAttach op has round-tripped.
-    private readonly pendingBlobIds: Map<string, Deferred<void>> = new Map();
-    // blobs uploaded while detached; cleared upon attach
+    private readonly pendingBlobIds: Map<string, Deferred<void> | undefined> = new Map();
+    // blob IDs assigned while detached; cleared upon attach
     private readonly detachedBlobIds: Set<string> = new Set();
-    // map of detached blob IDs to IDs used by storage. used to support blob handles given out while detached
-    private redirectTable: Map<string, string> | undefined;
+    // map of local blob IDs to IDs used by storage. used to support blob handles given out while detached or offline
+    private readonly redirectTable = new Map<string, string>();
+
+    // contains local IDs assigned while offline
+    private readonly offlineBlobs = new Set<string>();
+    private readonly offlineBlobStorage = new MockDetachedBlobStorage();
 
     constructor(
         private readonly routeContext: IFluidHandleContext,
         snapshot: IBlobManagerLoadInfo,
         private readonly getStorage: () => IDocumentStorageService,
-        private readonly attachBlobCallback: (blobId: string) => void,
+        private readonly sendBlobAttachOp: (blobId?: string, localId?: string) => void,
         private readonly runtime: IContainerRuntime,
         private readonly logger: ITelemetryLogger,
     ) {
         this.runtime.once("dispose", () => {
             for (const promise of this.pendingBlobIds.values()) {
-                promise.reject(new Error("runtime disposed while blobAttach op in flight"));
+                promise?.reject(new Error("runtime disposed while blobAttach op in flight"));
             }
         });
         this.load(snapshot);
     }
 
     private hasBlob(id: string): boolean {
-        return this.blobIds.has(id) || this.detachedBlobIds.has(id);
+        return this.blobIds.has(id) || this.detachedBlobIds.has(id) ||
+            this.offlineBlobs.has(id) || this.redirectTable.has(id);
     }
 
-    public async getBlob(blobId: string): Promise<IFluidHandle<ArrayBufferLike>> {
-        const storageId = this.redirectTable?.get(blobId) ?? blobId;
-        assert(this.hasBlob(storageId), 0x11f /* "requesting unknown blobs" */);
+    private async getBlob(blobId: string): Promise<ArrayBufferLike> {
+        if (this.offlineBlobs.has(blobId)) {
```

Comment:
Can this knowledge (of offline blobs, and redirectTable) be hidden in adapter class? I.e. where this class only knows about one storage, but there is an adapter implementation sitting between real storage and this class?
The only thing this class needs to know about such adapter is to pass to it redirect table on attachment / boot.

Basically I'm looking for ways to split this class into more smaller classes that have fewer responsibilities, such that it's easier to write tests / validate code and understand machinery.

Reply:
Maybe it will be hard to achieve (looking at things like applyStashedBlobAttachOp, but I'm still interesting in this question - do you see any opportunities to split this class into smaller, mostly independent chunks?